src/sys/stm_tests.ml

```diff
@@ -0,0 +1,242 @@
+    | File of {perm: int}
+
+  type state = filesys
+
+  type sut   = unit
+
+  let (/) = Filename.concat
+
+  let update_map_name map_name k v = Map_names.add k v (Map_names.remove k map_name)
+
+  let arb_cmd _s  =
+    let name_gen = Gen.(oneofl ["aaa" ; "bbb" ; "ccc" ; "ddd" ; "eee"]) in
+    let path_gen = Gen.map (fun path -> path) (Gen.list_size (Gen.int_bound 5) name_gen) in (* can be empty *)
+    let perm_gen = Gen.(oneofl [0o777]) in
+    QCheck.make ~print:show_cmd
+      Gen.(oneof
+            [
+                map (fun path -> File_exists path) path_gen ;
+                map3 (fun path new_dir_name perm -> Mkdir (path, new_dir_name, perm)) path_gen name_gen perm_gen;
+                map2 (fun path delete_dir_name -> Rmdir (path, delete_dir_name)) path_gen name_gen;
+                map (fun path -> Readdir path) path_gen;
+                map3 (fun path new_file_name perm -> Touch (path, new_file_name, perm)) path_gen name_gen perm_gen;
+            ])
+
+  let static_path = Sys.getcwd ()
+
+  let init_state  = Directory {perm = 0o777; fs_map = Map_names.empty}
+
+  let rec find_opt_model fs path =
+    match fs with
+    | File f ->
+      if path = []
+      then Some (File f)
+      else None
+    | Directory d ->
+        (match path with
+        | []       -> Some (Directory d)
+        | hd :: tl ->
+          (match Map_names.find_opt hd d.fs_map with
+          | None    -> None
+          | Some fs -> find_opt_model fs tl))
+
+  let mem_model fs path = find_opt_model fs path <> None
+
+  let rec mkdir_model fs path new_dir_name perm =
+    match fs with
+    | File _      -> fs
+    | Directory d ->
+      (match path with
+      | [] ->
+        let new_dir = Directory {perm; fs_map = Map_names.empty} in
+        Directory {d with fs_map = Map_names.add new_dir_name new_dir d.fs_map}
+      | next_in_path :: tl_path ->
+        (match Map_names.find_opt next_in_path d.fs_map with
+        | None        -> fs
+        | Some sub_fs ->
+          let nfs = mkdir_model sub_fs tl_path new_dir_name perm in
+          if nfs = sub_fs
+          then fs
+          else
+            let new_map = Map_names.remove next_in_path d.fs_map in
+            let new_map = Map_names.add next_in_path nfs new_map in
+            Directory {d with fs_map = new_map}))
+
+  let readdir_model fs path =
+    match find_opt_model fs path with
+    | None    -> None
+    | Some fs ->
+      match fs with
+      | File _ -> None
+      | Directory d -> Some (Map_names.fold (fun k _ l -> k::l) d.fs_map [])
+
+  let rec rmdir_model fs path delete_dir_name =
+    match fs with
+    | File _      -> fs
+    | Directory d ->
+      (match path with
+      | [] ->
+        (match Map_names.find_opt delete_dir_name d.fs_map with
+        | Some (Directory target) when Map_names.is_empty target.fs_map ->
+          Directory {d with fs_map = Map_names.remove delete_dir_name d.fs_map}
+        | None | Some (File _) | Some (Directory _) -> fs)
```

Comment:
Isn't this last `Some (Directory _)` case redundant?

Reply:
If I remove it, where do non_empty directories go in the pattern match ? 
In none because it is either Some ...WITH this condition
                                          or None or File..     ?

Reply:
Sorry, I overlooked the `when` clause! (I use them very seldomly myself)